packages/flutter/lib/src/foundation/assertions.dart

```diff
@@ -3,19 +3,114 @@
 // found in the LICENSE file.
 
 import 'basic_types.dart';
+import 'diagnostics.dart';
 import 'print.dart';
 
 /// Signature for [FlutterError.onError] handler.
 typedef FlutterExceptionHandler = void Function(FlutterErrorDetails details);
 
 /// Signature for [FlutterErrorDetails.informationCollector] callback
 /// and other callbacks that collect information into a string buffer.
-typedef InformationCollector = void Function(StringBuffer information);
+typedef InformationCollector = void Function(List<DiagnosticsNode> information);
+
+class _ErrorDiagnostic extends DiagnosticsProperty<List<Object>> {
+  _ErrorDiagnostic._fromParts(
+    List<Object> messageParts, {
+    DiagnosticsTreeStyle style = DiagnosticsTreeStyle.whitespace,
+    DiagnosticLevel level = DiagnosticLevel.info,
+  }) : assert(messageParts != null),
+       super(
+         null,
+         messageParts,
+         showName: false,
+         showSeparator: false,
+         defaultValue: null,
+         style: style,
+         level: level,
+       );
+
+  String valueToString({ TextTreeConfiguration parentConfiguration }) {
+    return value.join('');
+  }
+}
+
+/// Diagnostics describing an error message.
+///
+/// An error message should start with a summary that is typically
+/// no longer than 2 lines that concisely states what the assertion
+/// failure or contract violation was. The summary should be separated from the
+/// rest of the error message with a line break.
+///
+/// The client (e.g., an IDE) usually displays the error summary in red.
+///
+/// The full error message should include at least the following elements:
+/// * Claim: Explanation of the assertion failure or contract violation.
+/// * Grounds: Facts about the user's code that led to the error.
+/// * Warranty: Connections between the grounds and the claim.
+///
+/// Error messages may include references to objects relevant to the error using
+/// string interpolation and those objects will be directly available to
+/// debugging tools when error messages are generating as part of a debug builds.
+/// Other builds will see the same string error message but will not be able to
+/// extract the contents of the object.
+class ErrorMessage extends _ErrorDiagnostic {
+  // This constructor provides a reliable hook for a kernel transformer to find
+  // error messages that need to be rewritten to include object references.
+  //
+  // The message will display with the same text regardless of whether the
+  // kernel transformer is used but without the kernel transformer, debugging
+  // tools are unable to provide interactive displays of objects referenced by
+  // the string.
+  /// In debug builds, a kernel transformer rewrites calls to the default
+  /// constructor into calls to this constructor.
+  ///
+  /// of this class as follows so that the objects referenced in the
+  /// string literal can be captured by debuggers.
+  /// ```dart
```

Comment:
use the "## Sample code" or "{@tool}" magic to make sure these blocks get analyzed.

Reply:
I'll clean that up before sending for a real review. This CL is intended to just convey enough to make sure we are on the same page and I'm not going out and implementing something that the team is not bought into.

Reply:
(`## Sample code` is going away in https://github.com/flutter/flutter/pull/27793, but +1 to suing `{@tool ..}`)